### ISSUE TITLE
Embedded Terminal

### ISSUE THREAD

**Issue by mherrmann** [fman-community/fman#39]
This would have the huge advantage that you could easily navigate directories with fman, seeing what's inside them, yet at the same time you would be able to run commands.

Unfortunately, implementing a terminal emulator is a large task, and it will be difficult to do so well that it actually yields benefits over the many existing Terminal integrations / solutions.

Several users (in particular people who used FAR) have been very keen on this feature. I'm quite strongly against it, because it is a monumental task, will suck in comparison to "actual" terminals at the end of the day, and (imho) doesn't provide _that_ much added benefit over using F9 and F11 in fman with a little bit of Alt-Tab.

**Comment by bmix** [comment]
Maybe add a single text-input line, that executes manually entered shell commands in the active pane. Output would go to system's default terminal or could be caught by software allowing for it (ConEmu).

**Comment by vovcacik** [comment]
I never liked terminal integrated in file managers. But running commands on file/files is very useful - you may want to implement feature that will concat relative/absolute paths of selected files in fman and allow user to edit it to run arbitrary command. Something like `cmd /k "7z. a new.7z file1 file2 file3`.

**Comment by mherrmann** [comment]
Sorry, @vovcacik,  I don't understand the example you gave. One thing that's already there is `F11` - `CopyPathsToClipboard`. It copies the full path of the selected files to the clipboard. This can then be pasted into a terminal.

**Comment by vovcacik** [comment]
Ah I see, that would do the trick. My idea was that user would select files in fman, than he would be presented with simple input line (prompt) where paths of the files would be prefilled. Then he would edit the line to his needs - like adding `7z a` or whatever really. Once he confirm this dialog a normal terminal window would appear and carry out the command and stayed open once it finishes (to see the output).

But with the `F11` shortcut, I think this scenario is already pretty straightforward right now.

**Comment by mherrmann** [comment]
Ok good :)

**Comment by crispinb** [comment]
Completely agree on your reasoning for not implementing your own terminal. However, functionality-wise, the holy grail for terminal/file manager interactivity for me is the ability to have their cwds kept in sync so you can seamlessly jump between the two.  I haven't had a chance to look at your plugin API yet, but I'd suggest that's a place worth examining. It would be worth probing to see if there's anything missing that would enable a plugin to (minimally) with a quick command  or (preferably) automatically  keep the two in the same directory.

**Comment by raguay** [comment]
That is completely doable. When I get a chance, I'll add it to the iTerm
plugin. The problem facing is that it will be terminal program specific
since there is not generic way of doing it. But, people can create the
similar plugin for their favorite terminal program! Then, just load the one
you want to use.

**Comment by crispinb** [comment]
Yes it would need to be per-plugin. I just wasn't sure whether or not the current plugin API was rich enough to do this. If it is, then that's terrific.

**Comment by raguay** [comment]
Yes. There is a hook for when a directory changes. So, when a directory
changes and you want to sync, issue the command to the terminal program.
But, it will cost speed of fman as it will at least check for syncing every
time a directory changes.

**Comment by mherrmann** [comment]
As @raguay pointed out, it is possible to write an fman plugin that keeps a terminal in sync with regards to the current directory. I don't even think this will cause any noticeable slowdown in fman. The method one would have to overwrite is [DirectoryPaneListener#on_path_changed](https://fman.io/docs/api#DirectoryPaneListener.on_path_changed). This is invoked asynchronously by fman, ie. in the background.

**Comment by raguay** [comment]
I just updated OpenIterm2 plugin to allow for directory change following. Give it a try.

**Comment by crispinb** [comment]
@raguay thanks - I've been mac-free since July so I can't, but I'm glad others will benefit from your work! 
@mherrmann thanks for the info. I'm hoping to have a bit of time over the holidays to put together a plugin for my current os/terminal combination (Terminus on Win10), so that'll be helpful.

**Comment by raguay** [comment]
@crispinb Well, you can look at it on GitHub and see how to do it with your manager. So, are you windows or linux? I might be able to help you with Linux, but I don't have a working Windows system right now. My VM only have XP.
